osquery/tables/system/freebsd/xattr_utils.cpp

```diff
@@ -0,0 +1,181 @@
+  if (!s.ok()) {
+    return s;
+  }
+
+  size_t index = 0U;
+  while (index < raw_attribute_list.size()) {
+    auto name_length = static_cast<size_t>(raw_attribute_list[index]);
+    index++;
+
+    if (index + name_length > raw_attribute_list.size()) {
+      return Status(1, "Unexpected attribute list format");
+    }
+
+    auto current_attribute_name = raw_attribute_list.substr(index, name_length);
+    index += name_length;
+
+    attribute_list.push_back(current_attribute_name);
+  }
+
+  return Status(0);
+}
+
+Status getExtendedAttribute(std::string& value,
+                            const std::string& path,
+                            int namespace_,
+                            const std::string& name) {
+  value.clear();
+
+  auto buffer_size =
+      extattr_get_file(path.data(), namespace_, name.data(), nullptr, 0U);
+  if (buffer_size == -1) {
+    return Status(1, "Failed to retrieve the extended attribute length");
+  }
+
+  value.resize(static_cast<size_t>(buffer_size));
+  if (value.size() != static_cast<size_t>(buffer_size)) {
+    return Status(1, "Memory allocation failure");
+  }
+
+  buffer_size = extattr_get_file(
+      path.data(), namespace_, name.data(), &value[0], value.size());
+  if (buffer_size == -1) {
+    return Status(1, "Failed to read the extended attribute");
+  }
+
+  if (value.size() != static_cast<size_t>(buffer_size)) {
+    value.resize(buffer_size);
+  }
+
+  return Status(0);
+}
+
+Status appendExtendedAttributesForNamespace(ExtendedAttributes& attributes,
+                                            const std::string& path,
+                                            int namespace_) {
+  std::vector<std::string> attribute_list;
+  auto s = getExtendedAttributeList(attribute_list, path, namespace_);
+  if (!s.ok()) {
+    return s;
+  }
+
+  std::string name_prefix =
+      (namespace_ == EXTATTR_NAMESPACE_USER) ? "user." : "system.";
+
+  for (const auto& attribute_name : attribute_list) {
+    std::string attribute_value;
+    s = getExtendedAttribute(attribute_value, path, namespace_, attribute_name);
+    if (!s.ok()) {
+      return s;
+    }
+
+    attributes.push_back(
+        std::make_pair(name_prefix + attribute_name, attribute_value));
+  }
+
+  return Status(0);
+}
+} // namespace
+
+Status getExtendedAttributes(ExtendedAttributes& attributes,
+                             const std::string& path) {
+  attributes.clear();
+
+  for (int namespace_ : {EXTATTR_NAMESPACE_USER, EXTATTR_NAMESPACE_SYSTEM}) {
+    auto s = appendExtendedAttributesForNamespace(attributes, path, namespace_);
+    if (!s.ok()) {
+      return s;
+    }
+  }
+
+  return Status(0);
+}
+
+// Used by the tests in
+// osquery/tables/system/posix/tests/extended_attributes_tests.cpp
+Status setExtendedAttributes(
+    const std::string& path,
+    const std::unordered_map<std::string, std::string>& attributes) {
+  for (const auto& p : attributes) {
+    const auto& name = p.first;
+    const auto& value = p.second;
```

Comment:
Why create the extra local variables for the key and value of each hash map item?  Is this really needed?

Reply:
Without using those variables you have to track down the declaration to get the type and how it is used to understand what is returned.

The compiled output is the same whether you keep the variables or not, so it is more of a style decision; I have used them to make the code more readable (i.e. `expected.find(name)` vs `expected.find(p.first)`).

Reply:
Fair enough.  I would've thought that `attribute.first`/`attribute.second` were clear enough, but I definitely see where you are coming from.  :+1: